src/main.ts

```diff
@@ -0,0 +1,250 @@
+  stderr: SimpleWriteStream;
+  config: {
+    cachedRepositoriesDirectoryPath: string;
+    defaultProjectNames: string[];
+  };
+}) {
+  const outputLogger = new OutputLogger({ stdout, stderr });
+  const workingDirectoryPath = WORKING_DIRECTORY_PATH;
+
+  const inputs = await parseInputs({ argv, outputLogger, defaultProjectNames });
+  /* istanbul ignore next: At the moment, there is no real way that Yargs could fail */
+  if (!inputs) {
+    process.exitCode = 1;
+    return;
+  }
+  const { templateRepositoryName, projectReferences } = inputs;
+
+  const template = await establishMetaMaskRepository({
+    repositoryReference: templateRepositoryName,
+    workingDirectoryPath,
+    cachedRepositoriesDirectoryPath,
+    outputLogger,
+  });
+
+  await lintProjects({
+    projectReferences,
+    template,
+    workingDirectoryPath,
+    cachedRepositoriesDirectoryPath,
+    outputLogger,
+  });
+}
+
+/**
+ * Parses command-line arguments and establishes parameters for the remainder of
+ * this tool.
+ *
+ * @param args - The arguments to this function.
+ * @param args.argv - The name of this executable and its arguments (as obtained
+ * via `process.argv`).
+ * @param args.outputLogger - Writable streams for output messages.
+ * @param args.defaultProjectNames - The set of MetaMask repositories
+ * that will be linted.
+ * @returns The parsed command-line arguments if parsing was successful, or null
+ * otherwise.
+ */
+async function parseInputs({
+  argv,
+  outputLogger,
+  defaultProjectNames,
+}: {
+  argv: string[];
+  outputLogger: OutputLogger;
+  defaultProjectNames: string[];
+}): Promise<Inputs | null> {
+  const args = await parseCommandLineArguments({ argv, outputLogger });
+
+  if (args) {
+    const projectReferences = args._.length > 0 ? args._ : defaultProjectNames;
+    return {
+      templateRepositoryName: DEFAULT_TEMPLATE_REPOSITORY_NAME,
+      projectReferences,
+    };
+  }
+
+  /* istanbul ignore next: At the moment, there is no real way that Yargs could fail */
+  return null;
+}
+
+/**
+ * Parses the command line using `yargs`.
+ *
+ * @param args - The arguments to this function.
+ * @param args.argv - The name of this executable and its arguments (as obtained
+ * via `process.argv`).
+ * @param args.outputLogger - Writable streams for output messages.
+ */
+async function parseCommandLineArguments({
+  argv,
+  outputLogger,
+}: {
+  argv: string[];
+  outputLogger: OutputLogger;
+}) {
+  let yargsFailure: { message: string; error: Error } | null = null;
+  /* istanbul ignore next: At the moment, there is no real way that Yargs could fail */
+  const onFail = (message: string, error: Error) => {
+    if (error) {
+      throw error;
+    }
+    yargsFailure = { message, error };
+  };
+
+  const yargs = createYargs(hideBin(argv))
+    .usage(USAGE_TEXT)
+    .help(false)
+    .string('_')
+    .wrap(null)
+    .exitProcess(false)
+    .fail(onFail);
+
+  const options = await yargs.parse();
```

Comment:
I'm having trouble understanding the yargs docs on this method. How does this compare with using the `argv` like this:

```
const { argv } = createYargs(...)
```

Also my first time seeing this import called anything other than `yargs`; not sure I understand the choice there either.

Reply:
If you use `yargs` this way, the type of `argv` will be `{ ... } | Promise<{ ... }}>`. The reason for this is detailed in the [TypeScript docs](https://github.com/yargs/yargs/blob/main/docs/typescript.md), and the recommended solution is to await `argv`. So we could do:

``` typescript
const options = await yargs.argv;
```

That looks strange, so I figured I'd call `parse()` here, even if it does the same thing.

I also need the return value of the chain, because I need to call `.getHelp()` on it later. So I can't destructure it to pull `argv` out immediately, but I need to give that return value a name.

It seems that the Yargs documentation largely prefers requiring `yargs/yargs` directly, but in a few cases, e.g. the TypeScript docs they call the default import `yargs`, and then call the thing that `yargs` returns the `parser`. Then sometimes they call the `parser` `yargs`, e.g. for the [`command`](https://yargs.js.org/docs/#api-reference-commandcmd-desc-builder-handler) option. So it's a bit inconsistent, but I can settle on `parser` I suppose.

I've made these changes here. Hopefully this makes more sense: [1b5854d](https://github.com/MetaMask/module-lint/pull/33/commits/1b5854d392a65822afcf4f296dab7c648045ee73)